feat: pprof streaming parsing

korniltsev opened the pull request.

korniltsev commented on pkg/convert/pprof/decoder.go:
```diff
@@ -11,7 +11,7 @@ import (
 )
 
-var bufPool = bytebufferpool.Pool{}
+
```
I wanted to reuse a buffer with a current parser. I now made it to share one non global pool between them.

korniltsev commented on pkg/convert/pprof/streaming/parser_streaming.go:
```diff
@@ -0,0 +1,521 @@
+	p.startTime = startTime
+	p.endTime = endTime
+	p.ctx = ctx
+	defer func() { p.ctx = nil }()
```
it depends on how big is context retained size. there is no need to keep it. I rewrote without defer

korniltsev commented on pkg/convert/pprof/streaming/parser_streaming.go:
```diff
@@ -0,0 +1,521 @@
+			return err
+		}
+		p.profile = buf.Bytes()
+		defer func() { p.profile = nil }()
```
the field is to avoid passing the buffer around multiple functions , there is no need to keep profile after parsing is done since it may contain garbage. I remove a reference to it just to prevent possible access to the garbage in the future. I also rewrite without defer. OK?

korniltsev commented on pkg/convert/pprof/streaming/parser_streaming.go:
```diff
@@ -0,0 +1,521 @@
+// step 2
+// - parse locations
+// - parse functions
+func (p *MoleculeParser) parseStructs2() error {
```
It is now parseFunctionsAndLocations.

korniltsev commented on pkg/convert/pprof/streaming/labels_cache.go:
```diff
@@ -0,0 +1,125 @@
+	return l
+}
+
+// CutLabel creates a copy of labels without label i.
+func CutLabel(labels Labels, i int) Labels {
+	c := make(Labels, 0, len(labels)-1)
+	for j, label := range labels {
+		if i != j {
+			c = append(c, copyLabel(label))
+		}
+	}
+	return c
+}
+
+func copyLabel(label label) label {
+	return label
+}
```
made it 
```
func CutLabel(labels Labels, i int) Labels {
	c := make(Labels, len(labels)-1)
	copy(c[:i], labels[:i])
	copy(c[i:], labels[i+1:])
	return c
}
```

ok?

korniltsev commented on pkg/convert/pprof/streaming/parser_streaming.go:
```diff
@@ -0,0 +1,521 @@
+
+func (p *MoleculeParser) parsePprofDecompressed() error {
+	var err error
+	p.strings = make([][]byte, 0, 256) // todo sane default? count? reuse?
```
re strings size: the whole parsing happens in 3 pass, now I count number of strings in first pass and allocate correct slice size in second

korniltsev replied:
re pool: done

korniltsev commented on pkg/convert/pprof/profile.go:
```diff
@@ -187,35 +190,46 @@ func (p *RawProfile) Parse(ctx context.Context, putter storage.Putter, _ storage
-			filter := p.parser.sampleTypesFilter
-			p.parser.sampleTypesFilter = func(s string) bool {
+			filter := p.parser.GetSampleTypesFilter()
+			p.parser.SetSampleTypesFilter(func(s string) bool {
```
I wen with extra argument to ParsePprof

kolesnikovae commented on pkg/convert/pprof/parser.go:
```diff
@@ -165,15 +170,23 @@ func (p *Parser) iterate(x *tree.Profile, fn func(vt *tree.ValueType, l tree.Lab
-			types = append(types, s.Type)
+		st := x.StringTable[s.Type]
+		if p.sampleTypesFilter != nil && p.sampleTypesFilter(st) {
+			if cumulative {
+				if p.sampleTypes[st].Cumulative {
+					indexes = append(indexes, i)
+					types = append(types, s.Type)
+				}
+			} else {
+				indexes = append(indexes, i)
+				types = append(types, s.Type)
+			}
```
```suggestion
			if !cumulative || (cumulative && p.sampleTypes[st].Cumulative) {
				indexes = append(indexes, i)
				types = append(types, s.Type)
			}
```

kolesnikovae commented on pkg/convert/pprof/parser.go:
```diff
@@ -207,6 +220,7 @@ func (p *Parser) readTrees(x *tree.Profile, c tree.LabelsCache, f tree.Finder) {
 		}
 		// Insert tree nodes.
 		for i, vi := range indexes {
+			_ = i
```
Must be leftover after debugging

kolesnikovae commented on pkg/convert/pprof/profile.go:
```diff
@@ -187,35 +190,34 @@ func (p *RawProfile) Parse(ctx context.Context, putter storage.Putter, _ storage
+		}
 
 		if p.PreviousProfile != nil {
-			// Ignore non-cumulative samples from the PreviousProfile
-			// to avoid duplicates: although, presence of PreviousProfile
-			// tells that there are cumulative sample types, it may also
-			// include regular ones.
```
Please leave the comment :) I fixed this place two times in past and therefore decided to put it here: it is the only case when we want to grab only samples of cumulative type (also, it would be nice to rename the argument to something like `cumulativeOnly` just for better readability)

korniltsev replied:
it is just the comment is kinda related to a boolean arg and it is just "true", it felt it is not clear what the comment is reffering to
I guess if lave the comment and do cumulativeOnly:=true/false it would be better

kolesnikovae commented on pkg/convert/pprof/streaming/profile_finder.go:
```diff
@@ -0,0 +1,117 @@
+	locationsMap map[uint64]*location
+}
+
+//go:noinline
+func (f *finder) FindLocation(id uint64) (*location, bool) {
+	if f.locationsMap == nil {
+		idx := id - 1
+		//if idx >= uint64(len(f.locations)) {
+		//	return nil, false
+		//}
+		return &f.locations[idx], true
+	}
+	l, ok := f.locationsMap[id]
+	return l, ok
+}
+
+//go:noinline
+func (f *finder) FindFunction(id uint64) (*function, bool) {
+	if f.functionsMap == nil {
+		if id == 0 {
+			fmt.Println("hui")
+		}
+		idx := id - 1
+		//if idx >= uint64(len(f.functions)) {
+		//	return nil, false
+		//}
+		return &f.functions[idx], true
+	}
+	ff, ok := f.functionsMap[id]
+	return ff, ok
+}
```
Why not to inline them?

kolesnikovae replied:
Also, please clean up the code from debug helpers

korniltsev replied:
noinline is also leftover, will remove

korniltsev commented:
@kolesnikovae ptal, I rewrote the parse with code generated from VT, it is a bit faster and the code is a bit more reusable (except for the Profile message) and also more verbose.
I also addressed codereview issues
The pr is still work in progress, there are bunch of things I want to do here
- cleanup and commit a bench + comparing test
- implement stack formatter for ruby's pprof


I'd like to do the testing before merging the pr, do you think I need to do smth else for testing or I can try now

kolesnikovae commented on pkg/convert/pprof/streaming/structs.go:
```diff
@@ -0,0 +1,114 @@
+	tmpStackLoc []uint64
+}
+
+func (s *sample) reset() {
+	// 64 is max pc for golang + speculative number of inlines
+	if s.tmpStack == nil {
+		s.tmpStack = make([][]byte, 0, 64+8)
+		s.tmpStackLoc = make([]uint64, 0, 64+8)
+		s.tmpValues = make([]int64, 0, 4)
+	} else {
+		s.tmpStack = s.tmpStack[:0]
+		s.tmpStackLoc = s.tmpStackLoc[:0]
+		s.tmpValues = s.tmpValues[:0]
+	}
+}
+
+func (s *sample) resetSample() {
+	s.tmpValues = s.tmpValues[:0]
+	s.tmpLabels = s.tmpLabels[:0]
+	s.tmpStack = s.tmpStack[:0]
+	s.tmpStackLoc = s.tmpStackLoc[:0]
+}
```
I'm not sure I understand why we need two separate functions. Also, shouldn't `reset()` reset labels as well?

korniltsev replied:
it is a leftover, thanks

kolesnikovae commented on pkg/convert/pprof/streaming/vt_function.go:
```diff
@@ -0,0 +1,155 @@
+)
+
+// revive:disable-next-line:cognitive-complexity,cyclomatic necessary complexity
+func (m *function) UnmarshalVT(dAtA []byte) error {
```
These unmarshal methods are very repetitive. Can we do anything with it?

korniltsev replied:
I don't think it is a problem for small structs like function and locations
For Profile I had 3 huge almost identical copies for 3 passes, I merged them to a single function with an "operation" parameter

kolesnikovae commented:
I think if unit tests are okay we can test it with real data in our own tenant and polish everything after

kolesnikovae commented:
One thing I'd like to clarify – do you plan to reuse parsers? My understanding is that we allocate bunch of various slices that we can reuse

kolesnikovae commented on pkg/convert/pprof/profile.go:
```diff
@@ -187,35 +191,41 @@ func (p *RawProfile) Parse(ctx context.Context, putter storage.Putter, _ storage
-				}
-				return sampleTypes[s].Cumulative
+		if p.StreamingParser {
+			parser := streaming.VTStreamingParserFromPool(streaming.ParserConfig{
+				SpyName:       md.SpyName,
+				Labels:        md.Key.Labels(),
+				Putter:        putter,
+				SampleTypes:   sampleTypes,
+				SkipExemplars: p.SkipExemplars,
+			})
+			p.parser = parser
+			if p.PoolStreamingParser {
+				defer func() {
+					parser.ReturnToPool()
+					p.parser = nil
+				}()
```
Use of `streaming.VTStreamingParserFromPool` should depend on `p.PoolStreamingParser` as far as I understand

korniltsev replied:
it is, it is in a nested if branch.
or i don't get the comment?

kolesnikovae replied:
I mean that a parser should be only taken from pool if `p.PoolStreamingParser` is true, else it should be created with `NewStreamingParser`. Otherwise we will pollute the pool because we never return to it, which may cause higher memory consumption and some extra pressure on GC. However, `sync.Pool` implementation is quite smart and the impact might be unnoticeable.

Also, it is a common practice (according to my observations) to explicitly call `Reset` before returning an item to the pool in order to release (and make them available to GC) objects that can't be reused.

Frankly, I don't think the current implementation is wrong or not optimal, but there are some things that you are used to and expect to see

kolesnikovae commented on pkg/convert/pprof/streaming/vt_sample.go:
```diff
@@ -0,0 +1,293 @@
+}
+
+// revive:disable-next-line:cognitive-complexity,cyclomatic necessary complexity
+func (s *sample) UnmarshalSampleVT(dAtA []byte, tmpLabel *label) error {
```
It does not seem we modify it here (we should not at least as far as I can see) so please consider passing it by value - this is way cheaper to pass label by value

kolesnikovae replied:
I mean, we do not need a tmpLabel at all: it is better to allocate small types like below on the stack in the function body: 

```go

 type line struct {
 	functionID uint64
 }

 type label struct {
 	k, v int64
 }

```

kolesnikovae commented on pkg/convert/pprof/streaming/labels_cache.go:
```diff
@@ -0,0 +1,116 @@
+	"sort"
+)
+
+type Labels []label
```
Consider representing a label as a single uint64: we know for sure that both K and V << 2^32 so we can pack them `(k<<32)|v`. This reduces cost of:
1. sort
2. hashing <- I'm not even sure if we would need this: comparing underlying byte slices (of Labels) might be way cheaper

korniltsev replied:
Nice one.
I thought about another possible optimization: representing string as a uint64(pair of uint32, [from, to)) indices into profile buffer

kolesnikovae replied:
This should be also helpful. However, we still need to allocate a string for a reference (copy these bytes), otherwise memory ownership will be violated since the bytes are taken from pool. Therefore, depending on how many times a particular reference is resolved, it might even deteriorate the overall performance – though, I don't have a full picture and may be missing something

kolesnikovae commented on pkg/convert/pprof/streaming/labels_cache.go:
```diff
@@ -0,0 +1,116 @@
+
+// sample type -> labels hash -> entry
+
+type LabelsCache map[int64]map[uint64]*LabelsCacheEntry
```
In future we may want to optimise the structure – this is the hottest path.

We know that there aren't this many sample types and the set is dense, therefore we don't need the top level map, it should be replaces with a slice
```go
[]map[uint64]*LabelsCacheEntry
```

Next, `*LabelsCacheEntry` is actually a tree and associated Labels. We can represent them in Structure of Arrays way:

```go
type cache struct {
   // sample type -> labels hash -> index to labels and trees
   indices []map[uint64]int
   labels  []Labels
   trees   []*Tree
}
```

Another possible improvement is getting rid of Labels in favour of an arena:

```go
type cache struct {
   // sample type -> labels hash -> index to labelRefs and trees
   indices []map[uint64]int
   // A label reference points to the subset in labels:
   // Hight 32 bits is the start offset, lower 32 bits is the subset size.
   labelRefs []uint64
   labels    []Label
   trees     []*Tree
}
```

This allows reusing the cache easily. Also, `map[int]int can be optimised further (e.g https://github.com/brentp/intintmap).

korniltsev replied:
after our call I had an impression that with a new cloudstorage's ingesters we will not need the LabelsCache at all so I did not even try to optimize it

kolesnikovae replied:
Honestly, I'm not 100% sure when (if) it will be released as part of the OSS version. Also, the cache is only needed to do diffs of cumulative profiles, so I think we can't remove it completely :/ Moving the logic to the storage is kinda problematic, though possible I guess

korniltsev commented:
Actually let me try to implement these optimizations before the testing
